src/main/java/org/wildfly/security/authz/MappedRoleMapper.java

```
@@ -0,0 +1,131 @@
+ * limitations under the License.
+ */
+package org.wildfly.security.authz;
+
+import java.util.LinkedHashMap;
+import java.util.LinkedHashSet;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * A simple mapping role mapper.
+ *
+ * Maps each role to a set of new roles using a String to Set<String> map.
+ *
+ * @author <a href="mailto:mmazanek@redhat.com">Martin Mazanek</a>
+ */
+public class MappedRoleMapper implements RoleMapper {
+
+    private final Map<String, Set<String>> reverseRoleMap;
+
+    /**
+     * Construct a new instance.
+     * Called from WildFly core when using this as a custom component. You should not use this constructor and use {@link MappedRoleMapper.Builder} instead.
+     *
+     * @see MappedRoleMapper.Builder
+     */
+    public MappedRoleMapper() {
+        this.reverseRoleMap = new LinkedHashMap<>();
+    }
+
+
+    private MappedRoleMapper(Map<String, Set<String>> roleMap) {
+        this.reverseRoleMap = new LinkedHashMap<>();
+
+        Set<Map.Entry<String, Set<String>>> entrySet = roleMap.entrySet();
+
+        for (Map.Entry<String, Set<String>> entry : entrySet) {
+
+            for (String mappedRole : entry.getValue()) {
+                Set<String> rolesToMappedRole = reverseRoleMap.get(mappedRole);
+
+                if (rolesToMappedRole == null) {
+                    rolesToMappedRole = new LinkedHashSet<>();
+                    reverseRoleMap.put(mappedRole, rolesToMappedRole);
+                }
+
+                rolesToMappedRole.add(entry.getKey());
+            }
+        }
+    }
+
+    /**
+     * Custom component method.
+     * Called from WildFly core. Used to include mapped role mapping functionality in older WildFly versions.
+     *
+     * @param configuration map of mapping rules where key is delegate role and value is whitespace separated list of new roles
+     */
+    public void initialize(final Map<String, String> configuration) {
+        reverseRoleMap.clear();
```

Sorry, I has forgotten the object should be thread-safe.
Would be better to add volatile boolean, which would set the mapper read-only during initialization.
Reapeated initialization should throw an exception, as additional map initialization would be thread unsafe.
(Also in the non-parametrized constructor should be defined `initialize()` has to be called before the constructed object is provided into multi-threaded environment.